File: programmers/DFS&BFS/단어변환/단어변환.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

int check[100];
int a = 100;
bool oneChange(string a, string b) {
    int _count = 0;
    for(int i=0 ; i<a.size(); i++) {
        if(a.at(i) != b.at(i)) _count += 1;
    }
    if(_count == 1) return true;
    else return false;
}
void dfs(int n, string b, string t, vector<string> w) {
    if(b.compare(t) == 0) {
        a = min(a, n);
        return;
    }
    for(int i=0 ; i<w.size() ; i++) {
        if(!oneChange(b, w[i])) continue;
        if(check[i]==0) {
            check[i] = 1;
            dfs(n+1, w[i], t, w);
            check[i] = 0;
        }
    }
}
int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    
    if(find(words.begin(), words.end(), target) == words.end()) return 0;
    else {
        dfs(0, begin, target, words);
        return a;
    }
}

File: programmers/DFS&BFS/단어변환/단어변환.py
check = []
min_cnt = 100
def can_str(a, b):
    cnt = 0
    for i in range(len(a)):
        if a[i] != b[i]:
            cnt += 1
    return cnt == 1
def dfs(words, cur, target, cnt):
    global min_cnt
    if cur == target:
        min_cnt = min(min_cnt, cnt)
        return
    
    for i in range(len(words)):
        if check[i] == 1:
            continue
        if not can_str(cur, words[i]):
            continue
        check[i] = 1
        dfs(words, words[i], target, cnt+1)
        check[i] = 0

def solution(begin, target, words):
    answer = 0
    global check
    global min_cnt
    check = [0 for _ in range(len(words))]
    dfs(words, begin, target, 0)
    return 0 if (min_cnt == 100) else min_cnt